In [139]:
# Import neccessary libaries

import pandas as pd # Dataframes
import numpy as np # Arrays and Math

# Visual libraries
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

# Warnings set to Ignore messages
import warnings
warnings.filterwarnings('ignore')

In [10]:
# Import the sample dataset as a DataFrame
data = pd.read_csv('ted_main.csv')

In [11]:
# Create a copy DataFrame, keeping data as backup if needed
df = data.copy()

In [12]:
# View the top 5 rows of data
df.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  published_date  \
0      Ken Robinson: Do schools kill creativity?            1      1151367060   
1           Al Gore: Averting the climate crisis            1      1151367060   
2                  David Pogue: Simplicity sells            1      1151367060   
3             Majora Carter: Greening the ghetto            1      1151367060   
4  Hans Rosling: The best stats you've ever seen            1      1151440680   

                                             ratings  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520  
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292  
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550  
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869

In [13]:
# View the bottom 5 rows of data
df.tail()

comments                                        description  duration  \
2545        17  Between 2008 and 2016, the United States depor...       476   
2546         6  How can you study Mars without a spaceship? He...       290   
2547        10  Science fiction visions of the future show us ...       651   
2548        32  In an unmissable talk about race and politics ...      1100   
2549         8  With more than half of the world population li...       519   

             event   film_date  languages         main_speaker  \
2545       TED2017  1496707200          4     Duarte Geraldino   
2546       TED2017  1492992000          3  Armando Azua-Bustos   
2547       TED2017  1492992000          1       Radhika Nagpal   
2548  TEDxMileHigh  1499472000          1     Theo E.J. Wilson   
2549       TED2017  1492992000          1    Karoliina Korppoo   

                                                   name  num_speaker  \
2545  Duarte Geraldino: What we're missing in the de...            1   
2546  Armando Azua-Bustos: The most Martian place on...            1   
2547  Radhika Nagpal: What intelligent machines can ...            1   
2548  Theo E.J. Wilson: A black man goes undercover ...            1   
2549  Karoliina Korppoo: How a video game might help...            1   

      published_date                                            ratings  \
2545      1505851216  [{'id': 3, 'name': 'Courageous', 'count': 24},...   
2546      1505919737  [{'id': 22, 'name': 'Fascinating', 'count': 32...   
2547      1506006095  [{'id': 1, 'name': 'Beautiful', 'count': 14}, ...   
2548      1506024042  [{'id': 11, 'name': 'Longwinded', 'count': 3},...   
2549      1506092422  [{'id': 21, 'name': 'Unconvincing', 'count': 2...   

                                          related_talks   speaker_occupation  \
2545  [{'id': 2596, 'hero': 'https://pe.tedcdn.com/i...           Journalist   
2546  [{'id': 2491, 'hero': 'https://pe.tedcdn.com/i...       Astrobiologist   
2547  [{'id': 2346, 'hero': 'https://pe.tedcdn.com/i...    Robotics engineer   
2548  [{'id': 2512, 'hero': 'https://pe.tedcdn.com/i...  Public intellectual   
2549  [{'id': 2682, 'hero': 'https://pe.tedcdn.com/i...        Game designer   

                                                   tags  \
2545  ['TED Residency', 'United States', 'community'...   
2546  ['Mars', 'South America', 'TED Fellows', 'astr...   
2547  ['AI', 'ants', 'fish', 'future', 'innovation',...   
2548  ['Internet', 'TEDx', 'United States', 'communi...   
2549  ['cities', 'design', 'future', 'infrastructure...   

                                                  title  \
2545  What we're missing in the debate about immigra...   
2546                    The most Martian place on Earth   
2547  What intelligent machines can learn from a sch...   
2548       A black man goes undercover in the alt-right   
2549  How a video game might help us build better ci...   

                                                    url   views  
2545  https://www.ted.com/talks/duarte_geraldino_wha...  450430  
2546  https://www.ted.com/talks/armando_azua_bustos_...  417470  
2547  https://www.ted.com/talks/radhika_nagpal_what_...  375647  
2548  https://www.ted.com/talks/theo_e_j_wilson_a_bl...  419309  
2549  https://www.ted.com/talks/karoliina_korppoo_ho...  391721

In [14]:
# Inspect the overall shape of the data (rows/columns)
df.shape

(2550, 17)

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2550 entries, 0 to 2549
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   comments            2550 non-null   int64 
 1   description         2550 non-null   object
 2   duration            2550 non-null   int64 
 3   event               2550 non-null   object
 4   film_date           2550 non-null   int64 
 5   languages           2550 non-null   int64 
 6   main_speaker        2550 non-null   object
 7   name                2550 non-null   object
 8   num_speaker         2550 non-null   int64 
 9   published_date      2550 non-null   int64 
 10  ratings             2550 non-null   object
 11  related_talks       2550 non-null   object
 12  speaker_occupation  2544 non-null   object
 13  tags                2550 non-null   object
 14  title               2550 non-null   object
 15  url                 2550 non-null   object
 16  views               2550

The dataset is comprised of 2,550 rows and 17 columns.  There are **no null values** out of the 2,550 entries, however, Speaker Occupation appears to have 4 **duplicate values** which could be NAs and will require further analysis.

In [33]:
# Inspect total count of null values in dataset
df.isnull().sum()

comments              0
description           0
duration              0
event                 0
film_date             0
languages             0
main_speaker          0
name                  0
num_speaker           0
published_date        0
ratings               0
related_talks         0
speaker_occupation    6
tags                  0
title                 0
url                   0
views                 0
dtype: int64

In [35]:
# Inspect total count of NA (NaN) values in dataset
df.isna().sum()

comments              0
description           0
duration              0
event                 0
film_date             0
languages             0
main_speaker          0
name                  0
num_speaker           0
published_date        0
ratings               0
related_talks         0
speaker_occupation    6
tags                  0
title                 0
url                   0
views                 0
dtype: int64

Due to the fact that there are **only 6 speakers with missing occupations**, all of whom can likely be searched online for generic occupations, these entries can be **manually encoded**.

In [150]:
# Create a list of speakers with missing occupations
speakers = df['main_speaker'][df['speaker_occupation'].isna() == True].values.tolist()

# create an, ordered, input list for respective speaker occupations (from online search)
occupations_input = ['Designer', 'Entrepreneur', 'CEO', 'Inventor', 'Artist', 'Journalist']

# for loop to update each speaker with respective occupation that was missig
for speaker in speakers:
    for occupation in occupations_input:
        df['speaker_occupation'][df['main_speaker'] == speaker] = occupation

In [153]:
# Recheck the count of NAs to ensure above loop removed all
df.isna().sum()

comments              0
description           0
duration              0
event                 0
film_date             0
languages             0
main_speaker          0
name                  0
num_speaker           0
published_date        0
ratings               0
related_talks         0
speaker_occupation    0
tags                  0
title                 0
url                   0
views                 0
dtype: int64

In [154]:
# Run descriptive statistics summary for all variables - including categorical
df.describe(include = 'all').T

count unique  \
comments            2550.0    NaN   
description           2550   2550   
duration            2550.0    NaN   
event                 2550    355   
film_date           2550.0    NaN   
languages           2550.0    NaN   
main_speaker          2550   2156   
name                  2550   2550   
num_speaker         2550.0    NaN   
published_date      2550.0    NaN   
ratings               2550   2550   
related_talks         2550   2550   
speaker_occupation    2550   1458   
tags                  2550   2530   
title                 2550   2550   
url                   2550   2550   
views               2550.0    NaN   

                                                                  top freq  \
comments                                                          NaN  NaN   
description         Sir Ken Robinson makes an entertaining and pro...    1   
duration                                                          NaN  NaN   
event                                                         TED2014   84   
film_date                                                         NaN  NaN   
languages                                                         NaN  NaN   
main_speaker                                             Hans Rosling    9   
name                        Ken Robinson: Do schools kill creativity?    1   
num_speaker                                                       NaN  NaN   
published_date                                                    NaN  NaN   
ratings             [{'id': 7, 'name': 'Funny', 'count': 19645}, {...    1   
related_talks       [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...    1   
speaker_occupation                                             Writer   45   
tags                           ['live music', 'music', 'performance']    3   
title                                     Do schools kill creativity?    1   
url                 https://www.ted.com/talks/ken_robinson_says_sc...    1   
views                                                             NaN  NaN   

                                 mean               std           min  \
comments                   191.562353        282.315223           2.0   
description                       NaN               NaN           NaN   
duration                   826.510196        374.009138         135.0   
event                             NaN               NaN           NaN   
film_date           1321927552.047059  119739059.699232    74649600.0   
languages                   27.326275          9.563452           0.0   
main_speaker                      NaN               NaN           NaN   
name                              NaN               NaN           NaN   
num_speaker                  1.028235          0.207705           1.0   
published_date      1343525265.750588   94640090.798582  1151367060.0   
ratings                           NaN               NaN           NaN   
related_talks                     NaN               NaN           NaN   
speaker_occupation                NaN               NaN           NaN   
tags                              NaN               NaN           NaN   
title                             NaN               NaN           NaN   
url                               NaN               NaN           NaN   
views                  1698297.481569    2498479.372185       50443.0   

                             25%           50%            75%           max  
comments                    63.0         118.0         221.75        6404.0  
description                  NaN           NaN            NaN           NaN  
duration                   577.0         848.0        1046.75        5256.0  
event                        NaN           NaN            NaN           NaN  
film_date           1257465600.0  1333238400.0   1412964000.0  1503792000.0  
languages                   23.0          28.0           33.0          72.0  
main_speaker                 NaN           NaN            NaN           NaN  
name      

**Initial Sample Observations**
- The most frequest event (year) is TED2014 (84 occurences)
- Hans Rosling was the most frequent speaker (9 presentations)
- The most common speaker occupation is Writer (45 occurences)

In [23]:
# Inspect the top 10 speaker occupations
df['speaker_occupation'].value_counts().head(10)

Writer          45
Artist          34
Designer        34
Journalist      33
Entrepreneur    31
Architect       30
Inventor        27
Psychologist    26
Photographer    25
Filmmaker       21
Name: speaker_occupation, dtype: int64